dali/pipeline/operators/reader/nvdecoder/imgproc.cu

```diff
@@ -0,0 +1,140 @@
+#include "dali/pipeline/operators/reader/nvdecoder/imgproc.h"
+
+#include <cuda_fp16.h>
+
+namespace dali {
+
+namespace {
+
+// using math from https://msdn.microsoft.com/en-us/library/windows/desktop/dd206750(v=vs.85).aspx
+
+template<typename T>
+struct yuv {
+    T y, u, v;
+};
+
+__constant__ float yuv2rgb_mat[9] = {
+    1.164383f,  0.0f,       1.596027f,
+    1.164383f, -0.391762f, -0.812968f,
+    1.164383f,  2.017232f,  0.0f
+};
+
+__device__ float clip(float x, float max) {
+    return fmin(fmax(x, 0.0f), max);
+}
+
+template<typename T>
+__device__ T convert(const float x) {
+    return static_cast<T>(x);
+}
+
+template<>
+__device__ half convert<half>(const float x) {
+    return __float2half(x);
+}
+
+template<>
+__device__ uint8_t convert<uint8_t>(const float x) {
+    return static_cast<uint8_t>(roundf(x));
+}
+
+template<typename YUV_T, typename RGB_T>
+__device__ void yuv2rgb(const yuv<YUV_T>& yuv, RGB_T* rgb,
+                        size_t stride, bool normalized) {
+    auto mult = normalized ? 1.0f : 255.0f;
+    auto y = (static_cast<float>(yuv.y) - 16.0f/255) * mult;
+    auto u = (static_cast<float>(yuv.u) - 128.0f/255) * mult;
+    auto v = (static_cast<float>(yuv.v) - 128.0f/255) * mult;
+
+    auto& m = yuv2rgb_mat;
+
+    // could get tricky with a lambda, but this branch seems faster
+    float r, g, b;
+    if (normalized) {
+        r = clip(y*m[0] + u*m[1] + v*m[2], 1.0);
+        g = clip(y*m[3] + u*m[4] + v*m[5], 1.0);
+        b = clip(y*m[6] + u*m[7] + v*m[8], 1.0);
+    } else {
+        r = clip(y*m[0] + u*m[1] + v*m[2], 255.0);
+        g = clip(y*m[3] + u*m[4] + v*m[5], 255.0);
+        b = clip(y*m[6] + u*m[7] + v*m[8], 255.0);
+    }
+
+    rgb[0] = convert<RGB_T>(r);
+    rgb[stride] = convert<RGB_T>(g);
+    rgb[stride*2] = convert<RGB_T>(b);
+}
+
+template<typename T>
+__global__ void process_frame_kernel(
+    cudaTextureObject_t luma, cudaTextureObject_t chroma,
+    T* dst, int index,
+    float fx, float fy,
+    int dst_width, int dst_height, int c) {
+
+    const int dst_x = blockIdx.x * blockDim.x + threadIdx.x;
+    const int dst_y = blockIdx.y * blockDim.y + threadIdx.y;
+
+    if (dst_x >= dst_width || dst_y >= dst_height)
+        return;
+
+    auto src_x = 0.0f;
+    // if (dst.desc.horiz_flip) {
+    //     src_x = (dst.desc.scale_width - dst.desc.crop_x - dst_x) * fx;
+    // } else {
+    //     src_x = (dst.desc.crop_x + dst_x) * fx;
+    // }
+    src_x = static_cast<float>(dst_x) * fx;
+    // auto src_y = static_cast<float>(dst_y + dst.desc.crop_y) * fy;
+    auto src_y = static_cast<float>(dst_y) * fy;
+
+
+    // https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#tex2d-object
+    yuv<float> yuv;
+    yuv.y = tex2D<float>(luma, src_x + 0.5, src_y + 0.5);
+    auto uv = tex2D<float2>(chroma, (src_x / 2) + 0.5, (src_y / 2) + 0.5);
```

Comment:
Unless our decoder does some resampling of the original YCbCr content, then most likely it should be (src_y + 0.5f) * 0.5f (the chroma sample is located between luma lines - hence, pixel centre
As for src_x, I can see there are competing sampling modes in different codecs. MPEG-1 used chroma sample at the centre of the 2x2 luma block, MPEG2 uses vertical centre, but horizontal left edge (which makes your code correct). I don't know what happens for other codecs.

Reply:
Unless we confirm that this code work I would avoid this kind of changes - I mean that we train some network with it and it has the same accuracy.